packages/webviz-core/src/dataProviders/MemoryCacheDataProvider.js

```diff
@@ -268,19 +273,46 @@ export default class MemoryCacheDataProvider implements DataProvider {
     this._maybeRunNewConnections();
   }
 
+  _getNewConnection() {
+    const connectionForReadRange = getNewConnection({
+      currentRemainingRange: this._currentConnection ? this._currentConnection.remainingBlockRange : undefined,
+      readRequestRange: this._readRequests[0] ? this._readRequests[0].blockRange : undefined,
+      downloadedRanges: this._getDownloadedBlockRanges(),
+      lastResolvedCallbackEnd: this._lastResolvedCallbackEnd,
+      cacheSize: READ_AHEAD_BLOCKS,
+      fileSize: this._blocks.length,
+      continueDownloadingThreshold: 3, // Somewhat arbitrary number to not create new connections all the time.
+    });
+    if (connectionForReadRange) {
+      return connectionForReadRange;
+    }
+    const cacheBytesUsed = sum(filterMap(this._blocks, (block) => block && block.sizeInBytes));
+    if (!this._currentConnection && cacheBytesUsed < CACHE_SIZE_BYTES) {
+      // All read requests have been served, but we have free cache space available. Cache something
+      // useful if possible.
+      return this._getPrefetchRange();
+    }
+    // Either a good connection is already in progress, or we've served all connections and have
+    // nothing useful to prefetch.
+  }
+
+  _getPrefetchRange() {
+    const bounds = { start: 0, end: this._blocks.length };
+    const uncachedRanges = missingRanges(bounds, this._getDownloadedBlockRanges());
+    if (!uncachedRanges.length) {
+      return; // We have loaded the whole file.
+    }
+
+    const prefetchStart = getPrefetchStartPoint(uncachedRanges, this._lastResolvedCallbackEnd || 0);
+    return { start: prefetchStart, end: prefetchStart + 1 };
```

Comment:
Instead of `prefetchStart + 1`, should we set this to the start of the next downloaded range? It's a bit minor, but we'd get some small benefits from it, e.g. when the next read request comes in and it's about to get downloaded (e.g. it's only a few blocks "away"), we won't be terminating the connection (which can be expensive) per the `continueDownloadingThreshold`.

Reply:
The start of the next downloaded range could be far away, so I guess we need a way to cut off that connection if we hit our memory budget if we do that.

Also, it would cause blocks in `_recentBlockRanges` to merge. Is that problematic? Would we instead want to aim on block earlier?

Reply:
Ah yeah I just realised the memory budget issue. In that case I _think_ that the current approach is a bit clearer. Might be worth a comment somewhere?

Re blocks in `_recentBlockRanges` merging; that will also happen with the current approach right, when we fill in all the blocks in between two ranges? I think that's okay, and even expected.

Reply:
Yeah, we want it to happen if they *actually* merge, but if they're a million miles apart I think it's a bad outcome -- we'll never actually join them up until/unless we play the whole intervening interval, but we make the later block "seem recently used" right away.

The "one block short" hack isn't great either -- if it's preloading a big chunk of data and the user seeks somewhere else, there's a big invisible range inside `_recentBlockRanges` that could interfere with things, I think.

We could request a larger range and signal to `_setConnection` that this is a low-priority connection that can be stopped as soon as we start evicting things. It's another hack, though.

I've added a comment explaining why the code does what it does.